## Scenario - PA 3.1.5

Mandisa Nkosi is actively pursuing a career in data analysis. 
She has established a relationship with a political party that needs to decide 
how best to invest its available advertising budget. Mandisa thinks she can 
explore potential advertising avenues by analysing films that are available 
on streaming platforms. She will use the movies_merge.xlsx and ott_merge.csv data sets.

The insights gained from the analysis will inform the campaign, 
promotional materials, slogans, and language the political party will use 
to reach potential voters. You’ll return to this analysis throughout the week. 

## Objective 

To prepare your files and Jupyter Notebook for analysis:

    1. Import the CSV files into DataFrames.
    2. View the DataFrames.
    3. Describe the DataFrames to understand the structures and data types. 
    4. Merge the two DataFrames into a single DataFrame.

In [3]:
# Import the library.
import pandas as pd
import numpy as np

In [4]:
# Import the datasets.
movies_merge = pd.read_excel('movies_merge.xlsx')
ott_merge = pd.read_csv('ott_merge.csv')

In [5]:
# View the 'movies_merge' DataFrame.
print("Shape of the movies_merge df is, ",movies_merge.shape)
movies_merge.head()

Shape of the movies_merge df is,  (16744, 11)


ID                           Title  Year  Age  IMDb  Rotten Tomatoes  \
0   1                       Inception  2010  13+   8.8             0.87   
1   2                      The Matrix  1999  18+   8.7             0.87   
2   3          Avengers: Infinity War  2018  13+   8.5             0.84   
3   4              Back to the Future  1985   7+   8.5             0.96   
4   5  The Good, the Bad and the Ugly  1966  18+   8.8             0.97   

                        Directors                            Genres  \
0               Christopher Nolan  Action,Adventure,Sci-Fi,Thriller   
1  Lana Wachowski,Lilly Wachowski                     Action,Sci-Fi   
2         Anthony Russo,Joe Russo           Action,Adventure,Sci-Fi   
3                 Robert Zemeckis           Adventure,Comedy,Sci-Fi   
4                    Sergio Leone                           Western   

                        Country                 Language  Runtime  
0  United States,United Kingdom  English,Japanese,French    148.0  
1                 United States                  English    136.0  
2                 United States                  English    149.0  
3                 United States                  English    116.0  
4      Italy,Spain,West Germany                  Italian    161.0

In [6]:
# View the 'ott_merge' DataFrame.
print("Shape of the ott_merge df is, ",ott_merge.shape)
ott_merge.head()

Shape of the ott_merge df is,  (16744, 5)


ID  Netflix  Hulu  Prime Video  Disney+
0   1        0     0            1        0
1   2        0     1            0        0
2   3        0     0            1        0
3   4        1     0            0        0
4   5        0     0            1        0

In [7]:
# Describe the DataFrame
print("The data types for movies_merge are:\n",movies_merge.dtypes)
print("\nThe data types for ott_merge are:\n",ott_merge.dtypes)

The data types for movies_merge are:
 ID                   int64
Title               object
Year                 int64
Age                 object
IMDb               float64
Rotten Tomatoes    float64
Directors           object
Genres              object
Country             object
Language            object
Runtime            float64
dtype: object

The data types for ott_merge are:
 ID             int64
Netflix        int64
Hulu           int64
Prime Video    int64
Disney+        int64
dtype: object


In [8]:
# Merge the DataFrame.
merged_data = pd.merge(movies_merge, ott_merge,
                      on='ID', how='left')

# View the merged_data
print("Shape of the merged_data df is, ",merged_data.shape)
print("\nThe data types for merged_data are:\n",merged_data.dtypes)
merged_data.head()

Shape of the merged_data df is,  (16744, 15)

The data types for merged_data are:
 ID                   int64
Title               object
Year                 int64
Age                 object
IMDb               float64
Rotten Tomatoes    float64
Directors           object
Genres              object
Country             object
Language            object
Runtime            float64
Netflix              int64
Hulu                 int64
Prime Video          int64
Disney+              int64
dtype: object


ID                           Title  Year  Age  IMDb  Rotten Tomatoes  \
0   1                       Inception  2010  13+   8.8             0.87   
1   2                      The Matrix  1999  18+   8.7             0.87   
2   3          Avengers: Infinity War  2018  13+   8.5             0.84   
3   4              Back to the Future  1985   7+   8.5             0.96   
4   5  The Good, the Bad and the Ugly  1966  18+   8.8             0.97   

                        Directors                            Genres  \
0               Christopher Nolan  Action,Adventure,Sci-Fi,Thriller   
1  Lana Wachowski,Lilly Wachowski                     Action,Sci-Fi   
2         Anthony Russo,Joe Russo           Action,Adventure,Sci-Fi   
3                 Robert Zemeckis           Adventure,Comedy,Sci-Fi   
4                    Sergio Leone                           Western   

                        Country                 Language  Runtime  Netflix  \
0  United States,United Kingdom  English,Japanese,French    148.0        0   
1                 United States                  English    136.0        0   
2                 United States                  English    149.0        0   
3                 United States                  English    116.0        1   
4      Italy,Spain,West Germany                  Italian    161.0        0   

   Hulu  Prime Video  Disney+  
0     0            1        0  
1     1            0        0  
2     0            1        0  
3     0            0        0  
4     0            1        0

In [9]:
# Concatenate the DataFrame.
concat_data = pd.concat([movies_merge, ott_merge], axis=0)

# View the concat_data
print("Shape of the concat_data df is, ",concat_data.shape)
print("\nThe data types for merged_data are:\n",concat_data.dtypes)
concat_data.head()

Shape of the concat_data df is,  (33488, 15)

The data types for merged_data are:
 ID                   int64
Title               object
Year               float64
Age                 object
IMDb               float64
Rotten Tomatoes    float64
Directors           object
Genres              object
Country             object
Language            object
Runtime            float64
Netflix            float64
Hulu               float64
Prime Video        float64
Disney+            float64
dtype: object


ID                           Title    Year  Age  IMDb  Rotten Tomatoes  \
0   1                       Inception  2010.0  13+   8.8             0.87   
1   2                      The Matrix  1999.0  18+   8.7             0.87   
2   3          Avengers: Infinity War  2018.0  13+   8.5             0.84   
3   4              Back to the Future  1985.0   7+   8.5             0.96   
4   5  The Good, the Bad and the Ugly  1966.0  18+   8.8             0.97   

                        Directors                            Genres  \
0               Christopher Nolan  Action,Adventure,Sci-Fi,Thriller   
1  Lana Wachowski,Lilly Wachowski                     Action,Sci-Fi   
2         Anthony Russo,Joe Russo           Action,Adventure,Sci-Fi   
3                 Robert Zemeckis           Adventure,Comedy,Sci-Fi   
4                    Sergio Leone                           Western   

                        Country                 Language  Runtime  Netflix  \
0  United States,United Kingdom  English,Japanese,French    148.0      NaN   
1                 United States                  English    136.0      NaN   
2                 United States                  English    149.0      NaN   
3                 United States                  English    116.0      NaN   
4      Italy,Spain,West Germany                  Italian    161.0      NaN   

   Hulu  Prime Video  Disney+  
0   NaN          NaN      NaN  
1   NaN          NaN      NaN  
2   NaN          NaN      NaN  
3   NaN          NaN      NaN  
4   NaN          NaN      NaN

When the concat function was used the duplicates were included in the final DataFrame,
so the total rows are 33484. But the merge funtion takes care of the duplicates and
so the total rows of the DataFrame is 16744. 

### Objective - PA 3.1.9 Use groupby() and aggregate() functions
Use the groupby() and aggregate() functions on the merged DataFrame to:

    1. Determine how many films from each year (released from 2012 to the present) 
       were watched on Netflix.
    2. Determine what the average runtime is of movies released each year.
    3. Determine what the best and worst reviews are that movies received 
       on Rotten Tomatoes.

### 1. Determine how many films from each year (released from 2012 to the present) were watched on Netflix. 

In [10]:
mo_gpby = merged_data.groupby('Year')[['Netflix']].agg('sum').reset_index()
mo_gpby[mo_gpby['Year']>=2012]

Year  Netflix
100  2012      195
101  2013      200
102  2014      202
103  2015      252
104  2016      236
105  2017      294
106  2018      307
107  2019      138
108  2020       31

### 2. Determine what the average runtime is of movies released each year since 2012.

In [11]:
mo_gpby = merged_data.groupby('Year')[['Runtime']].agg(['sum','mean']).reset_index()
mo_gpby[mo_gpby['Year']>=2012]

Year   Runtime           
                sum       mean
100  2012   73405.0  90.511714
101  2013   85687.0  91.545940
102  2014   87788.0  92.995763
103  2015   95795.0  92.914646
104  2016  107621.0  93.992140
105  2017  125822.0  94.460961
106  2018  112995.0  94.635678
107  2019   60997.0  93.410413
108  2020   12029.0  93.976562

### 3. Determine what the best and worst reviews are that movies received on Rotten Tomatoes since 2012.

In [12]:
mo_gpby = merged_data.groupby('Year')[['Rotten Tomatoes']].agg(['max','min']).reset_index()
mo_gpby[mo_gpby['Year']>=2012]

Year Rotten Tomatoes      
                      max   min
100  2012             1.0  0.04
101  2013             1.0  0.02
102  2014             1.0  0.05
103  2015             1.0  0.05
104  2016             1.0  0.02
105  2017             1.0  0.04
106  2018             1.0  0.06
107  2019             1.0  0.05
108  2020             1.0  0.06

## Scenario

The political party is considering running advertisements with some provoking messages,
but they need to ensure the content is suitable for the audiences watching the films. 
The data Mandisa is working with includes content ratings (in the Age column) 
that Mandisa can use to organise the DataFrame.

## Objective

Your objective at this stage is to organise the data by:

    -the film release date and content rating
    -the title of movies, the directors, and genres by content rating
    -the title of movies, the released year, and the language by content rating
    -Netflix screened movies based on language, runtime, and country
    -the title of movies, specified language, potential runtime, and genres by content rating.

You’ll need to use the pivot() function to create pivot tables for each description. 
More specifically, you’ll:

    -create a DataFrame for each description
    -employ the pivot() function on the merged and original DataFrames
    -specify the necessary parameters for the pivot() function.

In [13]:
# the film release date and content rating.
merged_data.pivot(index='Title', columns='Age', values='Year')

Age                                NaN     13+  16+     18+      7+  all
Title                                                                   
Inception                          NaN  2010.0  NaN     NaN     NaN  NaN
The Matrix                         NaN     NaN  NaN  1999.0     NaN  NaN
Avengers: Infinity War             NaN  2018.0  NaN     NaN     NaN  NaN
Back to the Future                 NaN     NaN  NaN     NaN  1985.0  NaN
The Good, the Bad and the Ugly     NaN     NaN  NaN  1966.0     NaN  NaN
...                                ...     ...  ...     ...     ...  ...
The Ghosts of Buxley Hall       1980.0     NaN  NaN     NaN     NaN  NaN
The Poof Point                     NaN     NaN  NaN     NaN  2001.0  NaN
Sharks of Lost Island           2013.0     NaN  NaN     NaN     NaN  NaN
Man Among Cheetahs              2017.0     NaN  NaN     NaN     NaN  NaN
In Beaver Valley                1950.0     NaN  NaN     NaN     NaN  NaN

[16744 rows x 6 columns]

In [14]:
# the title of movies, the directors, and genres by content rating.
merged_data.pivot(index='Title', columns='Age', values=['Directors', 'Genres'])

Directors                           \
Age                                              NaN                      13+   
Title                                                                           
Inception                                        NaN        Christopher Nolan   
The Matrix                                       NaN                      NaN   
Avengers: Infinity War                           NaN  Anthony Russo,Joe Russo   
Back to the Future                               NaN                      NaN   
The Good, the Bad and the Ugly                   NaN                      NaN   
...                                              ...                      ...   
The Ghosts of Buxley Hall               Bruce Bilson                      NaN   
The Poof Point                                   NaN                      NaN   
Sharks of Lost Island                   Neil Gelinas                      NaN   
Man Among Cheetahs              Richard Slater-Jones                      NaN   
In Beaver Valley                         James Algar                      NaN   

                                                                     \
Age                             16+                             18+   
Title                                                                 
Inception                       NaN                             NaN   
The Matrix                      NaN  Lana Wachowski,Lilly Wachowski   
Avengers: Infinity War          NaN                             NaN   
Back to the Future              NaN                             NaN   
The Good, the Bad and the Ugly  NaN                    Sergio Leone   
...                             ...                             ...   
The Ghosts of Buxley Hall       NaN                             NaN   
The Poof Point                  NaN                             NaN   
Sharks of Lost Island           NaN                             NaN   
Man Among Cheetahs              NaN                             NaN   
In Beaver Valley                NaN                             NaN   

                                                      \
Age                                          7+  all   
Title                                                  
Inception                                   NaN  NaN   
The Matrix                                  NaN  NaN   
Avengers: Infinity War                      NaN  NaN   
Back to the Future              Robert Zemeckis  NaN   
The Good, the Bad and the Ugly              NaN  NaN   
...                                         ...  ...   
The Ghosts of Buxley Hall                   NaN  NaN   
The Poof Point                      Neal Israel  NaN   
Sharks of Lost Island                       NaN  NaN   
Man Among Cheetahs                          NaN  NaN   
In Beaver Valley                            NaN  NaN   

                                                      Genres  \
Age                                                      NaN   
Title                                                          
Inception                                                NaN   
The Matrix                                               NaN   
Avengers: Infinity War                                   NaN   
Back to the Future                                       NaN   
The Good, the Bad and the Ugly                           NaN   
...                                                      ...   
The Ghosts of Buxley Hall       Comedy,Family,Fantasy,Horror   
The Poof Point                                           NaN   
Sharks of Lost Island                            Documentary   
Man Among Cheetahs                               Documentary   
In Beaver Valley                    Documentary,Short,Family   

                                                                       \
Age                                                          13+  16+   
Title                                                                   
Inception   

In [15]:
# the title of movies, the released year, and the language by content rating
merged_data.pivot(index='Title', columns='Age', values=['Year', 'Language'])

Year                             Language  \
Age                              NaN   13+  16+   18+    7+  all      NaN   
Title                                                                       
Inception                        NaN  2010  NaN   NaN   NaN  NaN      NaN   
The Matrix                       NaN   NaN  NaN  1999   NaN  NaN      NaN   
Avengers: Infinity War           NaN  2018  NaN   NaN   NaN  NaN      NaN   
Back to the Future               NaN   NaN  NaN   NaN  1985  NaN      NaN   
The Good, the Bad and the Ugly   NaN   NaN  NaN  1966   NaN  NaN      NaN   
...                              ...   ...  ...   ...   ...  ...      ...   
The Ghosts of Buxley Hall       1980   NaN  NaN   NaN   NaN  NaN  English   
The Poof Point                   NaN   NaN  NaN   NaN  2001  NaN      NaN   
Sharks of Lost Island           2013   NaN  NaN   NaN   NaN  NaN  English   
Man Among Cheetahs              2017   NaN  NaN   NaN   NaN  NaN  English   
In Beaver Valley                1950   NaN  NaN   NaN   NaN  NaN  English   

                                                                       \
Age                                                 13+  16+      18+   
Title                                                                   
Inception                       English,Japanese,French  NaN      NaN   
The Matrix                                          NaN  NaN  English   
Avengers: Infinity War                          English  NaN      NaN   
Back to the Future                                  NaN  NaN      NaN   
The Good, the Bad and the Ugly                      NaN  NaN  Italian   
...                                                 ...  ...      ...   
The Ghosts of Buxley Hall                           NaN  NaN      NaN   
The Poof Point                                      NaN  NaN      NaN   
Sharks of Lost Island                               NaN  NaN      NaN   
Man Among Cheetahs                                  NaN  NaN      NaN   
In Beaver Valley                                    NaN  NaN      NaN   

                                              
Age                                  7+  all  
Title                                         
Inception                           NaN  NaN  
The Matrix                          NaN  NaN  
Avengers: Infinity War              NaN  NaN  
Back to the Future              English  NaN  
The Good, the Bad and the Ugly      NaN  NaN  
...                                 ...  ...  
The Ghosts of Buxley Hall           NaN  NaN  
The Poof Point                  English  NaN  
Sharks of Lost Island               NaN  NaN  
Man Among Cheetahs                  NaN  NaN  
In Beaver Valley                    NaN  NaN  

[16744 rows x 12 columns]

In [16]:
# Netflix screened movies based on language, runtime, and country.
merged_data.pivot(index='Title', columns='Netflix', values=['Language', 'Runtime', 'Country'])

Language          Runtime  \
Netflix                                               0        1       0   
Title                                                                      
Inception                       English,Japanese,French      NaN   148.0   
The Matrix                                      English      NaN   136.0   
Avengers: Infinity War                          English      NaN   149.0   
Back to the Future                                  NaN  English     NaN   
The Good, the Bad and the Ugly                  Italian      NaN   161.0   
...                                                 ...      ...     ...   
The Ghosts of Buxley Hall                       English      NaN   120.0   
The Poof Point                                  English      NaN    90.0   
Sharks of Lost Island                           English      NaN     NaN   
Man Among Cheetahs                              English      NaN     NaN   
In Beaver Valley                                    NaN  English     NaN   

                                                            Country  \
Netflix                             1                             0   
Title                                                                 
Inception                         NaN  United States,United Kingdom   
The Matrix                        NaN                 United States   
Avengers: Infinity War            NaN                 United States   
Back to the Future              116.0                           NaN   
The Good, the Bad and the Ugly    NaN      Italy,Spain,West Germany   
...                               ...                           ...   
The Ghosts of Buxley Hall         NaN                 United States   
The Poof Point                    NaN                 United States   
Sharks of Lost Island             NaN                 United States   
Man Among Cheetahs                NaN                 United States   
In Beaver Valley                 32.0                           NaN   

                                               
Netflix                                     1  
Title                                          
Inception                                 NaN  
The Matrix                                NaN  
Avengers: Infinity War                    NaN  
Back to the Future              United States  
The Good, the Bad and the Ugly            NaN  
...                                       ...  
The Ghosts of Buxley Hall                 NaN  
The Poof Point                            NaN  
Sharks of Lost Island                     NaN  
Man Among Cheetahs                        NaN  
In Beaver Valley                United States  

[16744 rows x 6 columns]

In [17]:
# the title of movies, specified language, potential runtime, and genres by content rating.
merged_data.pivot(index='Title', columns='Age', values=['Language', 'Runtime', 'Genres'])

Language                                \
Age                                 NaN                      13+  16+   
Title                                                                   
Inception                           NaN  English,Japanese,French  NaN   
The Matrix                          NaN                      NaN  NaN   
Avengers: Infinity War              NaN                  English  NaN   
Back to the Future                  NaN                      NaN  NaN   
The Good, the Bad and the Ugly      NaN                      NaN  NaN   
...                                 ...                      ...  ...   
The Ghosts of Buxley Hall       English                      NaN  NaN   
The Poof Point                      NaN                      NaN  NaN   
Sharks of Lost Island           English                      NaN  NaN   
Man Among Cheetahs              English                      NaN  NaN   
In Beaver Valley                English                      NaN  NaN   

                                                      Runtime              \
Age                                 18+       7+  all     NaN    13+  16+   
Title                                                                       
Inception                           NaN      NaN  NaN     NaN  148.0  NaN   
The Matrix                      English      NaN  NaN     NaN    NaN  NaN   
Avengers: Infinity War              NaN      NaN  NaN     NaN  149.0  NaN   
Back to the Future                  NaN  English  NaN     NaN    NaN  NaN   
The Good, the Bad and the Ugly  Italian      NaN  NaN     NaN    NaN  NaN   
...                                 ...      ...  ...     ...    ...  ...   
The Ghosts of Buxley Hall           NaN      NaN  NaN   120.0    NaN  NaN   
The Poof Point                      NaN  English  NaN     NaN    NaN  NaN   
Sharks of Lost Island               NaN      NaN  NaN     NaN    NaN  NaN   
Man Among Cheetahs                  NaN      NaN  NaN     NaN    NaN  NaN   
In Beaver Valley                    NaN      NaN  NaN    32.0    NaN  NaN   

                                                   \
Age                               18+     7+  all   
Title                                               
Inception                         NaN    NaN  NaN   
The Matrix                      136.0    NaN  NaN   
Avengers: Infinity War            NaN    NaN  NaN   
Back to the Future                NaN  116.0  NaN   
The Good, the Bad and the Ugly  161.0    NaN  NaN   
...                               ...    ...  ...   
The Ghosts of Buxley Hall         NaN    NaN  NaN   
The Poof Point                    NaN   90.0  NaN   
Sharks of Lost Island             NaN    NaN  NaN   
Man Among Cheetahs                NaN    NaN  NaN   
In Beaver Valley                  NaN    NaN  NaN   

                                                      Genres  \
Age                                                      NaN   
Title                                                          
Inception                                                NaN   
The Matrix                                               NaN   
Avengers: Infinity War                                   NaN   
Back to the Future                                       NaN   
The Good, the Bad and the Ugly                           NaN   
...                                                      ...   
The Ghosts of Buxley Hall       Comedy,Family,Fantasy,Horror   
The Poof Point                                           NaN   
Sharks of Lost Island                            Documentary   
Man Among Cheetahs                               Documentary   
In Beaver Valley                    Documentary,Short,Family   

                                                                       \
Age                                                          13+  16+   
Title                                                                   
Inception                       Action,Adventure,Sci-Fi,Thriller  NaN 

### 3.2.4 Practical activity: apply() function
## Scenario

The political party is now ready to buy advertising slots and has received 
quotes from a broadcaster. The fee for running an ad is a flat rate per 10 seconds
of advertisement, with a minimum purchase of 60 seconds of advertising time per film. 
The broadcaster also has a promotional, reduced fee for ads run during any documentary film.

## Objective
To help the political party decide how they might best invest their budget,
Mandisa will answer the following business questions:

    -What is the effect of adding 60 seconds (one minute) to each movie?
    -Which movies are documentaries?

Your objective is to answer the posed questions by using the apply() 
and/or applymap() functions depending on whether you’re working on a 
Pandas Series or a Pandas DataFrame.

#### Q1. What is the effect of adding 60 seconds (one minute) to each movie?

In [19]:
# Create a subset
merged_data_new = merged_data[['ID', 'Title', 'Genres', 'Runtime']]

# View subset
merged_data_new.head()

ID                           Title                            Genres  \
0   1                       Inception  Action,Adventure,Sci-Fi,Thriller   
1   2                      The Matrix                     Action,Sci-Fi   
2   3          Avengers: Infinity War           Action,Adventure,Sci-Fi   
3   4              Back to the Future           Adventure,Comedy,Sci-Fi   
4   5  The Good, the Bad and the Ugly                           Western   

   Runtime  
0    148.0  
1    136.0  
2    149.0  
3    116.0  
4    161.0

In [21]:
# Add 60 seconds / 1 minute to the runtime
merged_data_new.Runtime.add(1)

0        149.0
1        137.0
2        150.0
3        117.0
4        162.0
         ...  
16739    121.0
16740     91.0
16741      NaN
16742      NaN
16743     33.0
Name: Runtime, Length: 16744, dtype: float64

#### Q2. Which movies are documentaries?

In [22]:
# Create a column to specify if Documentary or Not Documenatary.
merged_data_new['Gen_doc'] = np.where(merged_data_new['Genres'].str.contains
                                      ('Documentary'),'Documentary', 'Not Documentary')

# View
merged_data_new

/var/folders/hy/2ywwznf163g22yzbdl6zchbw0000gn/T/ipykernel_48877/1589540137.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_data_new['Gen_doc'] = np.where(merged_data_new['Genres'].str.contains


ID                           Title  \
0          1                       Inception   
1          2                      The Matrix   
2          3          Avengers: Infinity War   
3          4              Back to the Future   
4          5  The Good, the Bad and the Ugly   
...      ...                             ...   
16739  16740       The Ghosts of Buxley Hall   
16740  16741                  The Poof Point   
16741  16742           Sharks of Lost Island   
16742  16743              Man Among Cheetahs   
16743  16744                In Beaver Valley   

                                 Genres  Runtime          Gen_doc  
0      Action,Adventure,Sci-Fi,Thriller    148.0  Not Documentary  
1                         Action,Sci-Fi    136.0  Not Documentary  
2               Action,Adventure,Sci-Fi    149.0  Not Documentary  
3               Adventure,Comedy,Sci-Fi    116.0  Not Documentary  
4                               Western    161.0  Not Documentary  
...                                 ...      ...              ...  
16739      Comedy,Family,Fantasy,Horror    120.0  Not Documentary  
16740              Comedy,Family,Sci-Fi     90.0  Not Documentary  
16741                       Documentary      NaN      Documentary  
16742                       Documentary      NaN      Documentary  
16743          Documentary,Short,Family     32.0      Documentary  

[16744 rows x 5 columns]

In [28]:
# Determine the number of characters.
number_of_chars = merged_data_new.Gen_doc.apply(len)

# Creating new column 'Gen_doc_len' in the dataframe.
merged_data_new["Gen_doc_len"] = number_of_chars

#View Dataframe.
merged_data_new


/var/folders/hy/2ywwznf163g22yzbdl6zchbw0000gn/T/ipykernel_48877/2972970449.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_data_new["Gen_doc_len"] = number_of_chars


ID                           Title  \
0          1                       Inception   
1          2                      The Matrix   
2          3          Avengers: Infinity War   
3          4              Back to the Future   
4          5  The Good, the Bad and the Ugly   
...      ...                             ...   
16739  16740       The Ghosts of Buxley Hall   
16740  16741                  The Poof Point   
16741  16742           Sharks of Lost Island   
16742  16743              Man Among Cheetahs   
16743  16744                In Beaver Valley   

                                 Genres  Runtime          Gen_doc  Gen_doc_len  
0      Action,Adventure,Sci-Fi,Thriller    148.0  Not Documentary           15  
1                         Action,Sci-Fi    136.0  Not Documentary           15  
2               Action,Adventure,Sci-Fi    149.0  Not Documentary           15  
3               Adventure,Comedy,Sci-Fi    116.0  Not Documentary           15  
4                               Western    161.0  Not Documentary           15  
...                                 ...      ...              ...          ...  
16739      Comedy,Family,Fantasy,Horror    120.0  Not Documentary           15  
16740              Comedy,Family,Sci-Fi     90.0  Not Documentary           15  
16741                       Documentary      NaN      Documentary           11  
16742                       Documentary      NaN      Documentary           11  
16743          Documentary,Short,Family     32.0      Documentary           11  

[16744 rows x 6 columns]

### Challenge: 
Think of a way to help the political party save money. 
Mandisa suggests subtracting 6 seconds (0.01 minutes) from the 
Runtime of all the movies, resulting in saving if they 
limit the number of adverts during a movie. 

In [29]:
# Create a Subset
merged_data_run_less = merged_data[['ID', 'Title', 'Genres', 'Runtime']]

# View
merged_data_run_less.head()

ID                           Title                            Genres  \
0   1                       Inception  Action,Adventure,Sci-Fi,Thriller   
1   2                      The Matrix                     Action,Sci-Fi   
2   3          Avengers: Infinity War           Action,Adventure,Sci-Fi   
3   4              Back to the Future           Adventure,Comedy,Sci-Fi   
4   5  The Good, the Bad and the Ugly                           Western   

   Runtime  
0    148.0  
1    136.0  
2    149.0  
3    116.0  
4    161.0

In [31]:
# Sub 6 seconds(0.01minutes) from the runtime
merged_data_run_less.Runtime.subtract(0.01)

0        147.99
1        135.99
2        148.99
3        115.99
4        160.99
          ...  
16739    119.99
16740     89.99
16741       NaN
16742       NaN
16743     31.99
Name: Runtime, Length: 16744, dtype: float64